### ISSUE TITLE
sails-redis performance issue

### ISSUE THREAD

**Issue by billyshena** [balderdashy/sails-redis#80]
Hello guys ! 
I'm currently doing some benchmark on my application and found some performance issues regarding to Redis.
I have a simple API endpoint (Notification.find().limit(20)) and around 1200 notification rows in both dbs (Redis + Postgresql)
When I run a GET request on that endpoint (GET /notification), Redis is 10x slower than Postgresql.
Postgresql: 26ms to return 20 notification objects (from 1200 rows total)
Redis: 430ms

Does anyone know why Redis is so slow in fetching data? 

Best,

**Comment by Ryanc1256** [comment]
@billyshena I think its because sails-redis has to go find how many keys you actually have in a specific model, so..

Keys are like "waterline:notification:1" or something like that for your model, I think the problem is that
we have to figure out how many notification models there are, and do a single get for each one we thing is there, then run it through processing and then limit that then spit out then values... whereas if you just do it with redis you don't do any processing of it, and you don't do a get for each Key in the db, whereas I'm guessing your doing something like "waterline:notification:*" since the * which will get all the models correctly. 

Hope this helps.

**Comment by billyshena** [comment]
Hi @Ryanc1256 , thanks for your quick answer !
I do have my keys that look like this: "waterline:notification:id:717"
I'm currently using Waterline to fetch the data in my Redis database (simple Model.find().limit(1) would take 200 ms, same if I set .limit(1000)).
Would you suggest to use a native query to fetch my data from Redis?

Best,

**Comment by Ryanc1256** [comment]
@billyshena it really depends on what you want to do... if you want to have a lot of "joins" and doing the ".populate("ModelName")" thing, i would say stay with waterline as that is pretty intense to do by your self, but if your just doing something simple by just doing the .find( name ) then it might be easier to do native, but then again like i said you lose a lot of flexibility.

Personally, in one of my projects i use the native instead of ORM but i only use it due to the this module doesn't support ttl which im hoping someone will add soon (dreaming)... 

but going "native" means you have to do so much manual stuff, like processing for example...

doing something like this in waterline...
id = 1

Model.find({ id: id }).limit( 1 ) 

would turn into something like

Native.get("waterline:model:" + id, function( err, response ){
if ( err ){
   doSomething(err);
} else {
 var result = JSON.parse( response );
//do extra stuff...
}
});


So that is like the easiest thing i can think of, for a native query, and the more infomation you use, the harder it gets to maintain...

I love using ORMs as it just makes everything super super easy...

if you love making things, you could always make an simple ORM for redis only, or something, that would streamline a "Native" thing, but then you have to make a ORM loader and all that, it just makes the project a bit longer to build and then you have to figure out how to make queries fast and such... which is pretty hard

**Comment by billyshena** [comment]
Well I use Redis for simple Model.find() but I find that strange that a simple query like Model.find().limit(1) would take Redis around 200 ms to return the data (Postgresql = 20 ms)
I think I'll try in native just to compare the performance.

**Comment by Ryanc1256** [comment]
@billyshena yeah, well if its simple queries like that by all means go a head, it may be simpler until we get a better select method...

Since this has come up a couple times, and you mentioning it.... I think its time to look at how we do it and my mind is ticking away on ways to get it on par with postgressql and all that, well it should be way faster as it does far less and redis is the fastest DB atm ( i think?? )

You have to know also that Postgresql does cache queries whereas redis doesn't do anything like that, so if it simple select queries postgresql probably will comparable speeds  due to caching, if you turn that off you might see that Redis will blow Postgresql out of the water, well from my experience anyway.

**Comment by billyshena** [comment]
@Ryanc1256 I've tried using .native (KEYS *) as you mentioned and it is like 5-10 times faster than using Waterline .find() method to query my Redis keys.
I think Waterline is doing something wrong with the .find() method.

1. Model.native(function(err, collection) -> collection.keys('*') -> takes around 50 ms (1200 keys)
2. Model.find().exec() -> takes around 500 ms (1200 keys)

Hope a maintainer of the repo can take a look at the sails-redis adapter.

Best,

**Comment by Ryanc1256** [comment]
@billyshena woop's totally forgot to reply...

Anywho, hes my reply.


So waterline/redis isn't doing anything wrong but its doing it the like sql does it which I think is the wrong way of doing it...

But lets say you have 1200 keys
what this adapter does, keep a list off all the keys in the database so you would have something like

"waterline:model:_indexes" which has a list... of all the indexes in it...

so it would map to

waterline:model:1
waterline:model:2
waterline:model:3
waterline:model:4
... etc up to 
waterline:model:1200

So every time you do a ".find" it has to do another query to get that list and do the original query...

Model.find()
that's why it takes like 100 times as slow... 

But if your finding with a non-key so you have your database model like


Model
    id: key
    name: text
 ... etc...

So when you do a query like ".find({name: "bob"})"

this is where it gets a lot slower...

So it has to grab every key then process it, to see if name is equal to "bob" 

And P.S I'm pretty much the only maintainer of this repo :p
But don't worry, I'm planning to make this far far far better... but I'm just thinking of everything and making sure everything is way way way faster... so it's almost as fast as a native find...

Without breaking thing's for people that already use this module.-

**Comment by acolin** [comment]
+1 I also came across this issue and would love to help on a solution, thanks!

**Comment by Ryanc1256** [comment]
@acolin I think its a re-write of the query system to be smarter at what it does to go find the keys, which is a big undertaking I think and since its a big change we also need make sure it doesn't break anything from the previous versions....

I also want to add ttl as well which is redis key expiration too... but for later...

I just need to find some time to big all this together...

Thinking of doing it here...
https://github.com/Ryanc1256/sails-redis

Was thinking about a full re-write but I have no time for that... so this will do for now.

**Comment by Ryanc1256** [comment]
Okay I have more of an idea down so lets put it down on paper...

So first step is move the connection.get to an better way, so we can get some debug information in that so we know what queries are running without putting console.log or sails.log everywhere...

Now the 4 crud methods are going to change too... ( later )

> this will make all the methods more streamline and more away from how mysql does it and move it to an newer redis crud system and faster...
     > primarily the find method.

**Comment by billyshena** [comment]
@Ryanc1256 That would be awesome, do you plan to release a new version of this adapter?

Best,

**Comment by Ryanc1256** [comment]
Well yeah it will be a new major release... I think i got to talk it over with the other guys, but if your keen to look at that url above :P

**Comment by acolin** [comment]
@Ryanc1256 cool, will take a look at your repo

**Comment by Ryanc1256** [comment]
@acolin cool, i think ill try to put that debug info in this weekend, when i have a bit of free time and then i can start moving on to the other more intensive moves after i know whats fully going on... p.s. probably could do with some help with that, since its probably a big thing, so any help is appreciated :P

**Comment by acolin** [comment]
@Ryanc1256 sure, let me know once you have a big picture for the next release, I'll be glad to help.

**Comment by odnarb** [comment]
@Ryanc1256 @mikermcneil Hi guys, are any of these performance issues still relevant?

**Comment by johnabrams7** [comment]
@acolin @Ryanc1256 @billyshena Hey everyone, thanks for the all the input on this issue so far!  We're moving all the redis cases to the main Sails repo (balderdashy/sails/issues) to keep a closer eye on things and boost exposure to the entire community.
